[FSDP] Allow different `optim_input` orders across ranks

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#78599 [FSDP] Allow different `optim_input` orders across ranks**
* #78784 [FSDP][Docs] Fix typo in `full_optim_state_dict()`

This enables the first argument to the optimizer constructor, `optim_input`, to have different orders across ranks, e.g. if the parameters in one parameter group are permuted. This requires modification to `full_optim_state_dict()`, `shard_full_optim_state_dict()`, and `scatter_full_optim_state_dict()`.

The high-level algorithmic change is that the state dicts are kept as being keyed by unflattened parameter name until after sharding/unsharding and flattening/unflattening and are rekeyed to be by parameter ID according to each rank's own `optim_input` only at the end.

Because this PR adds non-parameter-specific collectives to `full_optim_state_dict()`, it adds a `group=None` argument to the method to have a process group to default to when running those common collectives.

Commit:
[FSDP] Allow diff `optim_input` across ranks

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36798482).

Commit:
Update on "[FSDP] Allow diff `optim_input` across ranks"


Differential Revision: [D36798482](https://our.internmc.facebook.com/intern/diff/D36798482)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36798482).

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -126,17 +131,9 @@ def _communicate_optim_state(
     state = _ConsolidatedOptimState()
     tensor_state, zero_dim_tensor_state, non_tensor_state = \
         state.tensor_state, state.zero_dim_tensor_state, state.non_tensor_state
-    process_group = fsdp_module.process_group
```
We do not have good support for multiple process groups in the module hierarchy. We can readdress if there are such use cases.

`full_optim_state_dict()` now needs to perform a `broadcast_object_list()` (to communicate rank 0's `optim_input` order) and an `all_reduce()` (for validation). These two collectives require _some_ default process group not necessarily tied to one module or parameter. This is why I felt like it would be simpler to just assume a single process group for the entire method.

rohan-varma commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -243,23 +244,32 @@ def _init_nested_model(
                 {"params": model.param_group1(), "weight_decay": 0.9}
             ]
+        # Use a reversed parameter order for the optimizer input on odd ranks
+        if use_diff_optim_inputs and self.rank % 2 == 1:
```
Do we get a similar error as we've seen internally with this change, without fixing the optim_state_dict logic?

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -840,6 +774,56 @@ def _broadcast_unsharded_pos_dim_tensor_state(
     param_state[state_name] = unsharded_tensor
 
 
+def _rekey_sharded_optim_state_dict(
```
This is different than the public `rekey_optim_state_dict()` since it is rekeying from an internal (private) `_OptimStateKey` to parameter ID.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -39,32 +38,37 @@
     init_from_local_shards,
 )
 from torch.distributed.distributed_c10d import _get_default_group
+from torch.distributed.utils import (
```
Sorted imports.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3496,62 +3500,91 @@ def full_optim_state_dict(
+            group = None  # use default group
+        else:
+            if len(process_groups) > 1:
+                warnings.warn(
```
Refer to the comment above about multiple process group support.

rohan-varma replied:
Can it break in unexpected ways if this is the case? If so, we should probably just be safe and raise an error here?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3496,62 +3500,91 @@ def full_optim_state_dict(
+        dist.broadcast_object_list(obj_list, src=0, group=group)
+        r0_flat_param_id_to_optim_state_key = obj_list[0]
+
+        # Ensure that all ranks have at least the optimizer states needed by
```
I think this validation is important to prevent hangs. All ranks should agree that they can proceed into the upcoming `for` loop over rank 0's `r0_optim_state_key`s without any issues before doing so. Otherwise, one rank might error, causing the others to hang.

rohan-varma commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -243,23 +244,32 @@ def _init_nested_model(
         group=None,
         optim_class: Type[torch.optim.Optimizer] = torch.optim.Adam,
-        use_multiple_param_groups: bool = False,
+        use_multiple_param_groups: bool = False,  # unused
```
why is there a # unused comment, but it is used later?

Commit:
Update on "[FSDP] Allow different `optim_input` orders across ranks"


This enables the first argument to the optimizer constructor, `optim_input`, to have different orders across ranks, e.g. if the parameters in one parameter group are permuted. This requires modification to `full_optim_state_dict()`, `shard_full_optim_state_dict()`, and `scatter_full_optim_state_dict()`.

The high-level algorithmic change is that the state dicts are kept as being keyed by unflattened parameter name until after sharding/unsharding and flattening/unflattening and are rekeyed to be by parameter ID according to each rank's own `optim_input` only at the end.

Differential Revision: [D36798482](https://our.internmc.facebook.com/intern/diff/D36798482)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM! Let's test to ensure that it is fixed for the usecase we're looking at. Thanks so much for the quick fix!

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3496,62 +3500,91 @@ def full_optim_state_dict(
+            raise RuntimeError(
+                "FSDP currently requires each rank to have at least the "
+                "optimizer states needed by rank 0's optimizer but some ranks "
+                "are missing some of those states"
```
Update: The error now looks like:
```
RuntimeError: FSDP currently requires each rank to have at least the optimizer states needed by rank 0's optimizer but some ranks are missing some of those states
Rank 1 is missing states for the parameters: [('block2.2.weight', 'block2.2.bias_module0.bias', 'block2.2.bias_module1.bias')]
Rank 2 is missing states for the parameters: [('block2.2.weight', 'block2.2.bias_module0.bias', 'block2.2.bias_module1.bias')]
Rank 3 is missing states for the parameters: [('block2.2.weight', 'block2.2.bias_module0.bias', 'block2.2.bias_module1.bias')]
```

Commit:
Update on "[FSDP] Allow different `optim_input` orders across ranks"


This enables the first argument to the optimizer constructor, `optim_input`, to have different orders across ranks, e.g. if the parameters in one parameter group are permuted. This requires modification to `full_optim_state_dict()`, `shard_full_optim_state_dict()`, and `scatter_full_optim_state_dict()`.

The high-level algorithmic change is that the state dicts are kept as being keyed by unflattened parameter name until after sharding/unsharding and flattening/unflattening and are rekeyed to be by parameter ID according to each rank's own `optim_input` only at the end.

Differential Revision: [D36798482](https://our.internmc.facebook.com/intern/diff/D36798482)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36798482).

Commit:
Update on "[FSDP] Allow different `optim_input` orders across ranks"


This enables the first argument to the optimizer constructor, `optim_input`, to have different orders across ranks, e.g. if the parameters in one parameter group are permuted. This requires modification to `full_optim_state_dict()`, `shard_full_optim_state_dict()`, and `scatter_full_optim_state_dict()`.

The high-level algorithmic change is that the state dicts are kept as being keyed by unflattened parameter name until after sharding/unsharding and flattening/unflattening and are rekeyed to be by parameter ID according to each rank's own `optim_input` only at the end.

Differential Revision: [D36798482](https://our.internmc.facebook.com/intern/diff/D36798482)

[ghstack-poisoned]

awgu commented:
Backward compatibility error due to broken ops seems unrelated.

awgu commented:
@pytorchbot merge

zhaojuanmao reviewed (commented):
cc @fegin as sharded optimizer states are built on top of it
